[BUG] Build fail with doctest 2.5.0

Hi,

### Prerequisites

* [x] Put an X between the brackets on this line if you have done all of the following:
    * Read about bug reporting in general: https://rspamd.com/doc/faq.html#how-to-report-bugs-found-in-rspamd
    * Enabled relevant debugging logs: https://rspamd.com/doc/faq.html#how-to-debug-some-module-in-rspamd 
    * Checked the FAQs about Core files in case of fatal crash: https://rspamd.com/doc/faq.html#how-to-figure-out-why-rspamd-process-crashed
    * Tried ASAN package and obtained the ASAN report (if possible): https://rspamd.com/doc/faq.html#asan-builds 
    * Checked that your issue isn't already filed: https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+user%3Arspamd
    * Checked that there is not already an experimental package or master branch

**Describe the bug**


When bumping doctest from 2.4.12 to 2.5.0 in nixpkgs, build fail with:

```cpp
[ 76%] Building CXX object src/CMakeFiles/rspamd-server.dir/libmime/mime_string.cxx.o
In file included from /build/source/src/libmime/mime_string.cxx:18:
/nix/store/4j70hhbg5xmlfspb1pfbcw49glsrm1jj-doctest-2.5.0/include/doctest/doctest.h: In instantiation of 'static void doctest::detail::filldata<T, typename std::enable_if<((! doctest::detail::has_insertion_operator<T>::value) && doctest::detail::is_container<T>::value)>::type>::fill(std::ostream*, const T&) [with T = rspamd::mime::basic_mime_string<char>; std::ostream = std::basic_ostream<char>]':
/nix/store/4j70hhbg5xmlfspb1pfbcw49glsrm1jj-doctest-2.5.0/include/doctest/doctest.h:1086:22:   required from 'void doctest::detail::filloss(std::ostream*, const T&) [with T = rspamd::mime::basic_mime_string<char>; std::ostream = std::basic_ostream<char>]'
 1086 |     filldata<T>::fill(stream, in);
      |     ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
/nix/store/4j70hhbg5xmlfspb1pfbcw49glsrm1jj-doctest-2.5.0/include/doctest/doctest.h:1099:12:   required from 'doctest::String doctest::detail::toStream(const T&) [with T = rspamd::mime::basic_mime_string<char>]'
 1099 |     filloss(stream, in);
      |     ~~~~~~~^~~~~~~~~~~~
/nix/store/4j70hhbg5xmlfspb1pfbcw49glsrm1jj-doctest-2.5.0/include/doctest/doctest.h:1107:24:   required from 'static doctest::String doctest::detail::StringMakerBase<true>::convert(const T&) [with T = rspamd::mime::basic_mime_string<char>]'
 1107 |         return toStream(in);
      |                ~~~~~~~~^~~~
/nix/store/4j70hhbg5xmlfspb1pfbcw49glsrm1jj-doctest-2.5.0/include/doctest/doctest.h:1144:35:   required from 'doctest::String doctest::toString(const T&) [with T = rspamd::mime::basic_mime_string<char>; typename std::enable_if<(! detail::should_stringify_as_underlying_type<T>::value), bool>::type <anonymous> = true]'
 1144 |     return StringMaker<T>::convert(value);
      |            ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
/nix/store/4j70hhbg5xmlfspb1pfbcw49glsrm1jj-doctest-2.5.0/include/doctest/doctest.h:1290:13:   required from 'doctest::String doctest::detail::stringifyBinaryExpr(const L&, const char*, const R&) [with L = rspamd::mime::basic_mime_string<char>; R = char [1]]'
 1290 |     return (DOCTEST_STRINGIFY(lhs)) + op + (DOCTEST_STRINGIFY(rhs));
      |             ^
/nix/store/4j70hhbg5xmlfspb1pfbcw49glsrm1jj-doctest-2.5.0/include/doctest/doctest.h:2000:5:   required from 'decltype (((void)((declval<L>() == declval<R>())), (doctest::detail::Result)(<brace-enclosed initializer list>()))) doctest::detail::Expression_lhs<L>::operator==(R&&) [with R = const char (&)[1]; L = const rspamd::mime::basic_mime_string<char>&; decltype (((void)((declval<L>() == declval<R>())), (doctest::detail::Result)(<brace-enclosed initializer list>()))) = doctest::detail::Result]'
 2000 |     DOCTEST_DO_BINARY_EXPRESSION_COMPARISON(==, " == ", DOCTEST_CMP_EQ)
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/build/source/src/libmime/mime_string.cxx:31:4:   required from here
   31 |                         CHECK(st == "");
      |                                     ^~
/nix/store/4j70hhbg5xmlfspb1pfbcw49glsrm1jj-doctest-2.5.0/include/doctest/doctest.h:1277:32: error: passing 'const rspamd::mime::basic_mime_string<char>' as 'this' argument discards qualifiers [-fpermissive]
 1277 |         for (auto it = in.begin(); it != in.end(); ++it) {
      |                        ~~~~~~~~^~
In file included from /build/source/src/libmime/mime_string.cxx:19:
/build/source/src/libmime/mime_string.hxx:526:21: note:   in call to 'rspamd::mime::basic_mime_string<T, Allocator, Functor>::iterator rspamd::mime::basic_mime_string<T, Allocator, Functor>::begin() [with CharT = char; Allocator = std::allocator<char>; Functor = fu2::abi_400::detail::function<fu2::abi_400::detail::config<false, true, fu2::capacity_default>, fu2::abi_400::detail::property<true, false, int(int)> >; iterator = rspamd::mime::iterator<rspamd::mime::basic_mime_string<char>, false>]'
  526 |         inline auto begin() noexcept -> iterator
      |                     ^~~~~
/nix/store/4j70hhbg5xmlfspb1pfbcw49glsrm1jj-doctest-2.5.0/include/doctest/doctest.h:1277:48: error: passing 'const rspamd::mime::basic_mime_string<char>' as 'this' argument discards qualifiers [-fpermissive]
 1277 |         for (auto it = in.begin(); it != in.end(); ++it) {
      |                                          ~~~~~~^~
/build/source/src/libmime/mime_string.hxx:536:21: note:   in call to 'rspamd::mime::basic_mime_string<T, Allocator, Functor>::iterator rspamd::mime::basic_mime_string<T, Allocator, Functor>::end() [with CharT = char; Allocator = std::allocator<char>; Functor = fu2::abi_400::detail::function<fu2::abi_400::detail::config<false, true, fu2::capacity_default>, fu2::abi_400::detail::property<true, false, int(int)> >; iterator = rspamd::mime::iterator<rspamd::mime::basic_mime_string<char>, false>]'
  536 |         inline auto end() noexcept -> iterator
      |                     ^~~
/nix/store/4j70hhbg5xmlfspb1pfbcw49glsrm1jj-doctest-2.5.0/include/doctest/doctest.h:1278:31: error: passing 'const rspamd::mime::basic_mime_string<char>' as 'this' argument discards qualifiers [-fpermissive]
 1278 |             if (it != in.begin()) {
      |                       ~~~~~~~~^~
/build/source/src/libmime/mime_string.hxx:526:21: note:   in call to 'rspamd::mime::basic_mime_string<T, Allocator, Functor>::iterator rspamd::mime::basic_mime_string<T, Allocator, Functor>::begin() [with CharT = char; Allocator = std::allocator<char>; Functor = fu2::abi_400::detail::function<fu2::abi_400::detail::config<false, true, fu2::capacity_default>, fu2::abi_400::detail::property<true, false, int(int)> >; iterator = rspamd::mime::iterator<rspamd::mime::basic_mime_string<char>, false>]'
  526 |         inline auto begin() noexcept -> iterator
      |                     ^~~~~
make[2]: *** [src/CMakeFiles/rspamd-server.dir/build.make:2305: src/CMakeFiles/rspamd-server.dir/libmime/mime_string.cxx.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1848: src/CMakeFiles/rspamd-server.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
```

I already tried a few simple things, like adding another `inline auto begin() noexcept -> iterator` with more `const`, but it was not enough.


### Steps to Reproduce

1. build with doctest 2.5.0

**Expected behavior**

should build

### Versions

N/A

### Additional Information

N/A

I see the error

> ```sh
>  error: 'this' argument to member function 'begin' has type 'const rspamd::mime::basic_mime_string<>', but function is not marked const
> ```

on an attempt to update MacPorts:

```sh
:info:build /opt/local/include/doctest/doctest.h:1277:24: error: 'this' argument to member function 'begin' has type 'const rspamd::mime::basic_mime_string<>', but function is not marked const 
:info:build  1277 |         for (auto it = in.begin(); it != in.end(); ++it) {
:info:build       |                        ^~
:info:build /opt/local/include/doctest/doctest.h:1086:18: note: in instantiation of member function 'doctest::detail::filldata<rspamd::mime::basic_mime_string<>>::fill' requested here
:info:build  1086 |     filldata<T>::fill(stream, in);
:info:build       |                  ^
:info:build /opt/local/include/doctest/doctest.h:1099:5: note: in instantiation of function template specialization 'doctest::detail::filloss<rspamd::mime::basic_mime_string<>>' requested here
:info:build  1099 |     filloss(stream, in);
:info:build       |     ^
:info:build /opt/local/include/doctest/doctest.h:1107:16: note: in instantiation of function template specialization 'doctest::detail::toStream<rspamd::mime::basic_mime_string<>>' requested here
:info:build  1107 |         return toStream(in);
:info:build       |                ^
:info:build /opt/local/include/doctest/doctest.h:1144:28: note: in instantiation of function template specialization 'doctest::detail::StringMakerBase<true>::convert<rspamd::mime::basic_mime_string<>>' requested here
:info:build  1144 |     return StringMaker<T>::convert(value);
:info:build       |                            ^
:info:build /opt/local/include/doctest/doctest.h:1290:13: note: in instantiation of function template specialization 'doctest::toString<rspamd::mime::basic_mime_string<>, true>' requested here
:info:build  1290 |     return (DOCTEST_STRINGIFY(lhs)) + op + (DOCTEST_STRINGIFY(rhs));
:info:build       |             ^
:info:build /opt/local/include/doctest/doctest.h:1123:32: note: expanded from macro 'DOCTEST_STRINGIFY'
:info:build  1123 | #define DOCTEST_STRINGIFY(...) toString(__VA_ARGS__)
:info:build       |                                ^
```

env:
```sh
:debug:build CC_PRINT_OPTIONS='YES'
:debug:build CC_PRINT_OPTIONS_FILE='/opt/local/var/macports/build/rspamd-0a014fb2/work/.CC_PRINT_OPTIONS'
:debug:build CPATH='/opt/local/include'
:debug:build DEVELOPER_DIR='/Library/Developer/CommandLineTools'
:debug:build LIBRARY_PATH='/opt/local/lib'
:debug:build MACOSX_DEPLOYMENT_TARGET='15.0'
:debug:build MACPORTS_LEGACY_SUPPORT_DISABLED='1'
:debug:build SDKROOT='/Library/Developer/CommandLineTools/SDKs/MacOSX15.sdk'
```

--- a/src/libmime/mime_string.hxx
+++ b/src/libmime/mime_string.hxx
@@ -94,7 +94,7 @@ public:
 		return idx != it.idx;
 	}
 
-	iterator_base(difference_type index, Container *instance) noexcept
+	iterator_base(difference_type index, const Container *instance) noexcept
 		: idx(index), cont_instance(instance)
 	{
 	}
@@ -102,11 +102,6 @@ public:
 	iterator_base(const iterator_base &) noexcept = default;
 
 	iterator_base &operator=(const iterator_base &) noexcept = default;
-
-	Container *get_instance() const noexcept
-	{
-		return cont_instance;
-	}
 
 	codepoint_type get_value() const noexcept
 	{
@@ -118,7 +113,7 @@ public:
 
 protected:
 	difference_type idx;
-	Container *cont_instance = nullptr;
+	const Container *cont_instance = nullptr;
 
 protected:
 	void advance(difference_type n) noexcept
@@ -164,7 +159,7 @@ public:
 		return idx != it.idx;
 	}
 
-	iterator_base(difference_type index, Container *instance) noexcept
+	iterator_base(difference_type index, const Container *instance) noexcept
 		: idx(index), cont_instance(instance)
 	{
 	}
@@ -172,10 +167,6 @@ public:
 	iterator_base() noexcept = default;
 	iterator_base(const iterator_base &) noexcept = default;
 	iterator_base &operator=(const iterator_base &) noexcept = default;
-	Container *get_instance() const noexcept
-	{
-		return cont_instance;
-	}
 
 	value_type get_value() const noexcept
 	{
@@ -184,7 +175,7 @@ public:
 
 protected:
 	difference_type idx;
-	Container *cont_instance = nullptr;
+	const Container *cont_instance = nullptr;
 
 protected:
 	//! Advance the iterator n times (negative values allowed!)
@@ -203,23 +194,21 @@ protected:
 	}
 };
 
-template<typename Container, bool Raw>
-struct iterator;
-template<typename Container, bool Raw>
-struct const_iterator;
 
 template<typename Container, bool Raw = false>
 struct iterator : iterator_base<Container, Raw> {
-	iterator(typename iterator_base<Container, Raw>::difference_type index, Container *instance) noexcept
-		: iterator_base<Container, Raw>(index, instance)
+	using base = iterator_base<Container, Raw>;
+	using typename base::difference_type;
+	using typename base::reference_type;
+
+	iterator(difference_type index, const Container *instance) noexcept
+		: base(index, instance)
 	{
 	}
 	iterator() noexcept = default;
 	iterator(const iterator &) noexcept = default;
 
 	iterator &operator=(const iterator &) noexcept = default;
-	/* Disallow creating from const_iterator */
-	iterator(const const_iterator<Container, Raw> &) = delete;
 
 	/* Prefix */
 	iterator &operator++() noexcept
@@ -251,33 +240,33 @@ struct iterator : iterator_base<Container, Raw> {
 		return tmp;
 	}
 
-	iterator operator+(typename iterator_base<Container, Raw>::difference_type n) const noexcept
+	iterator operator+(difference_type n) const noexcept
 	{
 		iterator it{*this};
 		it.advance(n);
 		return it;
 	}
 
-	iterator &operator+=(typename iterator_base<Container, Raw>::difference_type n) noexcept
+	iterator &operator+=(difference_type n) noexcept
 	{
 		this->advance(n);
 		return *this;
 	}
 
-	iterator operator-(typename iterator_base<Container, Raw>::difference_type n) const noexcept
+	iterator operator-(difference_type n) const noexcept
 	{
 		iterator it{*this};
 		it.advance(-n);
 		return it;
 	}
 
-	iterator &operator-=(typename iterator_base<Container, Raw>::difference_type n) noexcept
+	iterator &operator-=(difference_type n) noexcept
 	{
 		this->advance(-n);
 		return *this;
 	}
 
-	typename iterator::reference_type operator*() const noexcept
+	reference_type operator*() const noexcept
 	{
 		return this->get_value();
 	}
@@ -523,22 +512,22 @@ public:
 	}
 
 	/* Iterators */
-	inline auto begin() noexcept -> iterator
+	inline auto begin() const noexcept -> iterator
 	{
 		return {0, this};
 	}
 
-	inline auto raw_begin() noexcept -> raw_iterator
+	inline auto raw_begin() const noexcept -> raw_iterator
 	{
 		return {0, this};
 	}
 
-	inline auto end() noexcept -> iterator
+	inline auto end() const noexcept -> iterator
 	{
 		return {(difference_type) size(), this};
 	}
 
-	inline auto raw_end() noexcept -> raw_iterator
+	inline auto raw_end() const noexcept -> raw_iterator
 	{
 		return {(difference_type) size(), this};
 	}

